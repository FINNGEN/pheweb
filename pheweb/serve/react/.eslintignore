src/components/Variant/Lava/lavaa.js