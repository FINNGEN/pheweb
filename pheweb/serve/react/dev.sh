node_modules/webpack/bin/webpack.js --config webpack.dev.js --openssl-legacy-provider --watch
