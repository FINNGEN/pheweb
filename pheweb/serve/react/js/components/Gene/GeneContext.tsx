import React, { createContext, useEffect, useState } from "react";
import { GenePhenotypes } from "./geneModel";
import { fatal } from "../../common/Utilities";
import { getGenePhenotypes } from "./geneAPI";

interface Props { children: React.ReactNode }

interface Parameter {
  readonly phenotype : string | null
  readonly gene : string
}

export interface GeneState {
  readonly gene : string
  readonly genePhenotype : GenePhenotypes.Data
  readonly selectedPhenotype : GenePhenotypes.Phenotype
}

export const GeneContext = createContext<Partial<GeneState>>({})


export const createParameter = (href : string = window.location.href, fail = fatal) : Parameter | never => {
  const phenotype = new URLSearchParams(href).get('phenotype');
  const match = href.match("\/gene\/([^\/]+)(\/pheno\/([^\/]+))?$")
  if(match != null) {
    const [ignore_1, gene, ignore_2, phenotype,] = match
    return { gene , phenotype }
  }
  return fatal(`cant parse url ${href}`);
}

const GeneContextProvider = (props : Props) => {
  const parameter : Parameter = createParameter()
  const [genePhenotype, setGenePhenotype] = useState<GenePhenotypes.Data| undefined>(undefined);
  const [selectedPhenotype, setSelectedPhenotype] = useState<GenePhenotypes.Phenotype| undefined>(undefined);

  useEffect(() => { getGenePhenotypes(parameter.gene,setGenePhenotype) },[]);
  useEffect(() => {
    if(genePhenotype){
      const search = genePhenotype.phenotypes.find(p => p.pheno.phenocode == parameter.phenotype)
      if(search == undefined && genePhenotype.phenotypes.length > 0){
        setSelectedPhenotype(genePhenotype.phenotypes[0]);
      } else {
        setSelectedPhenotype(search);
      }
    }
  },[genePhenotype]);
  return (<GeneContext.Provider value={{ gene : parameter.gene,  selectedPhenotype , genePhenotype }}>
    {props.children}
  </GeneContext.Provider>)
}

export default GeneContextProvider