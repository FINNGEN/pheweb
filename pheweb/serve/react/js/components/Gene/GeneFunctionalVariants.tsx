import React, { useEffect, useState } from "react";
import { ConfigurationWindow } from "../Configuration/configurationModel";
import { mustacheDiv } from "../../common/Utilities";
import { FunctionalVariants } from "./geneModel";
import { getGeneFunctionalVariants, getLossOfFunction } from "./geneAPI";
import { Column } from "react-table";
import { createTableColumns, geneFunctionalVariantTableColumns } from "../../common/tableColumn";
import DownloadTable, { DownloadTableProps } from "../../common/DownloadTable";
import loading from "../../common/Loading";
import { finEnrichmentLabel } from "../Finngen/gnomad";

const default_banner : string =`
<div class="row">
  <div class="col-md-10 col-lg-10 col-sm-10 col-xs-10">
    <h3>All loss of function and missense variants</h3>
  </div>
</div>
`
const default_empty: string =`                                                                                                              
    No functional or missense variants for {{gene}}                                                                                               
  `

declare let window: ConfigurationWindow;
const config = window?.config?.userInterface?.gene?.functionalVariants;
const banner: string = config?.banner || default_banner;
const empty: string = config?.empty || default_empty;

const tableColumns : Column<FunctionalVariants.ViewRow>[] = createTableColumns<FunctionalVariants.ViewRow>(config?.tableColumns) || (geneFunctionalVariantTableColumns as Column<FunctionalVariants.ViewRow>[])
const defaultSorted =[]

const tableProperties = {  defaultPageSize : 5
}
const reshapeRow = (r : FunctionalVariants.Row) : FunctionalVariants.ViewRow => {
  const rsids = r.rsids
  const alt = r.var.alt
  const chr = r.var.chr
  const pos = r.var.pos
  const ref = r.var.ref
  const most_severe = r.var.annotation.annot.most_severe.replace(/_/g, ' ').replace(' variant', '')
  const info = r.var.annotation.annot.INFO
  const maf = +r.var.annotation.annot.AF < 0.5 ? +r.var.annotation.annot.AF : 1 - +r.var.annotation.annot.AF
  const fin_enrichment = finEnrichmentLabel(r.var.annotation.gnomad)
  const significant_phenos = r.significant_phenos

  return { rsids , alt , chr , pos , ref , most_severe , info , maf , fin_enrichment , significant_phenos }
}
const dataToTableRows = (data : FunctionalVariants.Data) : FunctionalVariants.ViewRow[] => data.map(reshapeRow)

interface Props { gene : string}
const GeneFunctionalVariants = ({ gene} ) => {

  const [data, setData] = useState<FunctionalVariants.Data | null>(null);
  useEffect(() => {
    getGeneFunctionalVariants(gene,setData)
  },[]);

  const prop : DownloadTableProps<FunctionalVariants.Data, FunctionalVariants.ViewRow> = {
    tableData : data,
    dataToTableRows ,
    tableColumns ,
    tableProperties,
    defaultSorted
  }
  const context = { gene }

  let view;
  if (data == null){
    view = loading;
  } else if(dataToTableRows(data).length == 0){
    view = <div>
      {mustacheDiv(banner, context)}
      {mustacheDiv(empty, context)}

    </div>
  } else {
    view = <div>
      {mustacheDiv(banner, context)}
      <DownloadTable {...prop}/>
    </div>
  }
  return view;

}

export default GeneFunctionalVariants