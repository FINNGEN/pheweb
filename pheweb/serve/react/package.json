{
  "name": "web",
  "version": "1.0.0",
  "description": "",
  "main": "webpack.common.js",
  "dependencies": {
    "@reduxjs/toolkit": "^1.4.0",
    "@types/d3": "^5.7.2",
    "@types/es6-promise": "^3.3.0",
    "@types/handlebars": "^4.1.0",
    "@types/jquery": "^3.5.0",
    "@types/mustache": "^4.1.2",
    "@types/node": "^14.0.23",
    "@types/react": "^16.14.20",
    "@types/react-csv": "^1.1.1",
    "@types/react-dom": "^16.9.14",
    "@types/react-router-dom": "^5.1.5",
    "@types/react-table": "^7.0.19",
    "@types/react-test-renderer": "^16.9.3",
    "@types/react-tooltip": "^4.2.4",
    "@types/uuid": "^8.3.3",
    "bootstrap": "^4.6.1",
    "d3": "^6.2.0",
    "d3-tip": "^0.9.1",
    "d3-v6-tip": "^1.0.9",
    "dts-gen": "^0.6.0",
    "handlebars": "^4.7.7",
    "jquery": "^3.6.0",
    "lavaa": "github:FINNGEN/lavaa-react#1541a557e5f5ca926dca3ef3088df713b3da288d",
    "locuszoom": "^0.11.0",
    "match-sorter": "^1.8.1",
    "npm-install-all": "^1.1.6",
    "npm-watch": "^0.11.0",
    "plotly.js": "^1.56.0",
    "q": "^1.5.1",
    "react": "^16.14.0",
    "react-csv": "^2.0.3",
    "react-dom": "^16.14.0",
    "react-plotly.js": "^2.5.0",
    "react-router-dom": "^5.2.0",
    "react-scripts": "^3.3.0",
    "react-table": "^6.8.6",
    "react-table-v6": "^6.8.6",
    "react-tabs": "^3.1.1",
    "react-tooltip": "^4.2.21",
    "regenerator-runtime": "^0.13.9",
    "ts-loader": "^8.0.11",
    "typescript": "^3.9.6",
    "underscore": "^1.13.2",
    "uuid": "^8.3.2",
    "validator": "^13.1.17"
  },
  "devDependencies": {
    "@testing-library/react": "^12.1.2",
    "@types/enzyme": "^3.10.7",
    "@types/enzyme-adapter-react-16": "^1.0.6",
    "@types/q": "^1.5.4",
    "@typescript-eslint/eslint-plugin": "^3.8.0",
    "@typescript-eslint/parser": "^3.8.0",
    "css-loader": "^5.0.2",
    "enzyme": "^3.11.0",
    "enzyme-adapter-react-16": "^1.15.6",
    "enzyme-to-json": "^3.6.2",
    "prettier": "2.5.0",
    "react-redux": "^7.2.1",
    "react-test-renderer": "^16.13.1",
    "redux": "^4.0.5",
    "standard": "^14.3.4",
    "standardx": "^5.0.0",
    "style-loader": "^2.0.0",
    "typescript-plugin-css-modules": "^3.4.0"
  },
  "scripts": {
    "watch": "npm-watch",
    "test": "react-scripts test --env=jsdom",
    "start": "react-scripts start",
    "build": "react-scripts build && cp -R build/* ../static/",
    "eject": "react-scripts eject"
  },
  "author": "",
  "license": "ISC",
  "eslintConfig": {
    "rules": {
      "no-unused-vars": "off",
      "@typescript-eslint/no-unused-vars": "error"
    }
  },
  "standardx": {
    "parser": "@typescript-eslint/parser",
    "plugins": [
      "@typescript-eslint/eslint-plugin"
    ]
  },
  "standard": {
    "ignore": [
      "js/pheno.js",
      "js/components/Variant.js",
      "js/components/Pheno.js",
      "js/components/Chip.js",
      "js/components/Coding.js",
      "js/components/index.js",
      "js/components/Index.js",
      "js/components/LoF.js",
      "js/components/Pheno.js",
      "js/components/Variant.js",
      "js/tables.js"
    ]
  },
  "browserslist": {
    "production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  },
  "watch": {
    "build": {
      "patterns": [
        "src"
      ],
      "extensions": "js,jsx,tsx"
    }
  }
}
