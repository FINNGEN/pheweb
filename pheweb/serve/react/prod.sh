node_modules/webpack/bin/webpack.js --config webpack.prod.js
