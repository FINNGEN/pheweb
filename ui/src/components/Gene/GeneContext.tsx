import React, { createContext, useEffect, useState } from "react";
import { GeneParams, GenePhenotypes } from "./geneModel";
import { getGenePhenotypes } from "./geneAPI";
import {setPageTitle} from "../../common/commonUtilities";

interface Props {
  readonly  children: React.ReactNode
  readonly  params : GeneParams
}

interface Parameter {
  readonly phenotype : string | null
  readonly gene : string
}

export interface GeneState {
  readonly gene : string
  readonly genePhenotype : GenePhenotypes.Data
  readonly selectedPhenotype : GenePhenotypes.Phenotype
  readonly errorMessage : string
  selectedTab : number
  setSelectedTab: React.Dispatch<React.SetStateAction<number>>
}

export const GeneContext = createContext<Partial<GeneState>>({})

export const createTitle = (gene : string, phenotype : string|undefined) => typeof phenotype === 'string'? `gene ${gene} ${phenotype}`:`gene ${gene}`;

const GeneContextProvider = ({ params : { gene , phenotype }, children } : Props) => {
  const [genePhenotype, setGenePhenotype] = useState<GenePhenotypes.Data| undefined>(undefined);
  const [selectedPhenotype, setSelectedPhenotype] = useState<GenePhenotypes.Phenotype| undefined>(undefined);
  const [errorMessage, setErrorMessage] = useState<string| undefined>(undefined);
  const [selectedTab, setSelectedTab] = useState<number>(0)

  const title : string = createTitle(gene, phenotype);
  setPageTitle(title);

  useEffect(() => { getGenePhenotypes(gene,setGenePhenotype, setErrorMessage) },[gene, setGenePhenotype]);
  useEffect(() => {
    if(genePhenotype){
      const search = genePhenotype.phenotypes.find(p => p.pheno.phenocode === phenotype)
      if(search === undefined && genePhenotype.phenotypes.length > 0){
        setSelectedPhenotype(genePhenotype.phenotypes[0]);
      } else {
        setSelectedPhenotype(search);
      }
    }
  },[genePhenotype, phenotype]);

  return (<GeneContext.Provider value={{ gene,
                                         selectedPhenotype ,
                                         genePhenotype ,
                                         errorMessage,
                                         selectedTab,
                                         setSelectedTab }}>
    { children }
  </GeneContext.Provider>)
}

export default GeneContextProvider
