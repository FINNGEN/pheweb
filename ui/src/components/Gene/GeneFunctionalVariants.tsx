import React, { useContext, useEffect, useState } from "react";
import { ConfigurationWindow } from "../Configuration/configurationModel";
import {defaultEmptyArray, flatten, mustacheDiv} from '../../common/commonUtilities';
import { FunctionalVariants } from './geneModel';
import { getGeneFunctionalVariants } from "./geneAPI";
import { Column } from "react-table";
import {
  wordFilter,
  createTableColumns,
  geneFunctionalVariantTableColumns,
  finnGenPhenotypeSubsetValues,
} from '../../common/commonTableColumn';
import CommonDownloadTable, { DownloadTableProps } from "../../common/CommonDownloadTable";
import commonLoading from "../../common/CommonLoading";
import { finEnrichmentLabel } from "../Finngen/finngenGnomad";
import { GeneContext, GeneState } from "./GeneContext";
import ViewRow = FunctionalVariants.ViewRow;

const default_banner : string =`
<div class="row">
  <div class="col-md-10 col-lg-10 col-sm-10 col-xs-10">
    <h3>All loss of function and missense variants</h3>
  </div>
</div>
`
const default_empty: string =`                                                                                                              
    No functional or missense variants for {{gene}}                                                                                               
  `

declare let window: ConfigurationWindow;
const config = window?.config?.userInterface?.gene?.functionalVariants;
const banner: string = config?.banner || default_banner;
const empty: string = config?.empty || default_empty;

const tableColumns : Column<FunctionalVariants.ViewRow>[] = createTableColumns<FunctionalVariants.ViewRow>(config?.tableColumns) || (geneFunctionalVariantTableColumns as Column<FunctionalVariants.ViewRow>[])
const defaultSorted =[]

const tableProperties = {
  defaultPageSize : 5,
  defaultFilterMethod : wordFilter
}
const reshapeRow = (r : FunctionalVariants.Row) : FunctionalVariants.ViewRow => {
  const rsids = r.rsids
  const alt = r.var.alt
  const chrom = r.var.chr
  const pos = r.var.pos
  const ref = r.var.ref
  const most_severe = r.var.annotation.annot.most_severe.replace(/_/g, ' ').replace(' variant', '')
  const info = r.var.annotation.annot.INFO
  const maf = +r.var.annotation.annot.AF < 0.5 ? +r.var.annotation.annot.AF : 1 - +r.var.annotation.annot.AF
  const fin_enrichment = finEnrichmentLabel(r.var.annotation.gnomad)
  const significant_phenos = r.significant_phenos

  return { rsids , alt , chrom , pos , ref , most_severe , info , maf , fin_enrichment , significant_phenos }
}
const dataToTableRows = (data : FunctionalVariants.Data) : FunctionalVariants.ViewRow[] => data.map(reshapeRow);

/* Create a row for significant_phenos when
 * downloading.
 */
const tableRowToDownloadRow = (columns : ViewRow[]) => flatten(columns.map(x => defaultEmptyArray(finnGenPhenotypeSubsetValues<FunctionalVariants.SignificantPheno>(x.significant_phenos),[x]).map(p => {  return { ...x, ...p}})   ));

interface Props { }
const GeneFunctionalVariants = () => {
  const { gene } = useContext<Partial<GeneState>>(GeneContext);
  const [data, setData] = useState<FunctionalVariants.Data | null>(null);
  const filename = `${gene}_functional_variants.tsv`
  useEffect(() => {
    getGeneFunctionalVariants(gene,setData)
  },[gene, setData]);
  
  const prop : DownloadTableProps<FunctionalVariants.Data, FunctionalVariants.ViewRow> = {
    filename,
    tableData : data,
    dataToTableRows ,
    tableColumns ,
    tableProperties,
    defaultSorted,
    tableRowToDownloadRow
  }
  const context = { gene }

  let view;
  if (data == null){
    view = commonLoading;
  } else if(dataToTableRows(data).length === 0){
    view = <div>
      {mustacheDiv(banner, context)}
      {mustacheDiv(empty, context)}

    </div>
  } else {
    view = <div>
      {mustacheDiv(banner, context)}
      <CommonDownloadTable {...prop}/>
    </div>
  }
  return view;

}

export default GeneFunctionalVariants