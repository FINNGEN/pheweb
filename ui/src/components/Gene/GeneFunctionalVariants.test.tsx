/* eslint-env jest */
import React from 'react';
import renderer from 'react-test-renderer';
import { render, fireEvent , screen } from '@testing-library/react'
import GeneFunctionalVariants from './GeneFunctionalVariants';
import { GeneContext, GeneState } from "./GeneContext";
import selectedPhenotype from "./GeneFunctionalVariants.test.data/selectedPhenotype.json";

test('TODO fix these tests', () => {})

/*
test('render without context', () => {
  const component = <GeneFunctionalVariants/>;
  const tree = renderer.create(component).toJSON();
  expect(tree).toMatchSnapshot()
})


test('render with context', () => {
    const GeneContextProvider = (params : { children }) => {
	const gene = "APOE";
	const genePhenotype = {   "phenotypes": [], "region": { "chrom": 19,
								"end": 45009393,
								"start": 44805791
							      } }; // GenePhenotypes.Data
	const errorMessage = undefined;

	return (<GeneContext.Provider value={{ gene,
                                         selectedPhenotype ,
                                         genePhenotype ,
                                         errorMessage }}>
    { params.children }
		</GeneContext.Provider>)
    }
  const component =   <GeneContextProvider>
	<GeneFunctionalVariants/>
  </GeneContextProvider>;
  const tree = renderer.create(component).toJSON();
  expect(tree).toMatchSnapshot()
})
*/
