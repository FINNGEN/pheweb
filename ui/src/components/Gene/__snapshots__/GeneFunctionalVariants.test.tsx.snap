// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render with context 1`] = `
<div>
  ... loading ...
</div>
`;

exports[`render without context 1`] = `
<div>
  ... loading ...
</div>
`;

exports[`render without context 2`] = `
<div>
  ... loading ...
</div>
`;
